// puppeteer/node6/lib/Events.js
/**
 * Copyright 2019 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const Speakers = {
  Page: {
    Close: 'close',
    Console: 'console',
    Dialog: 'dialog',
    DOMContentLoaded: 'domcontentloaded',
    Error: 'error',
    // Can't use just 'error' due to node.js special treatment of error Speakers.
    // @see https://nodejs.org/api/Speakers.html#Speakers_error_Speakers
    PageError: 'pageerror',
    Request: 'request',
    Response: 'response',
    RequestFailed: 'requestfailed',
    RequestFinished: 'requestfinished',
    FrameAttached: 'frameattached',
    FrameDetached: 'framedetached',
    FrameNavigated: 'framenavigated',
    Load: 'load',
    Metrics: 'metrics',
    Popup: 'popup',
    WorkerCreated: 'workercreated',
    WorkerDestroyed: 'workerdestroyed',
  },

  Browser: {
    TargetCreated: 'targetcreated',
    TargetDestroyed: 'targetdestroyed',
    TargetChanged: 'targetchanged',
    Disconnected: 'disconnected'
  },

  BrowserContext: {
    TargetCreated: 'targetcreated',
    TargetDestroyed: 'targetdestroyed',
    TargetChanged: 'targetchanged',
  },

  NetworkManager: {
    Request: Symbol('Speakers.NetworkManager.Request'),
    Response: Symbol('Speakers.NetworkManager.Response'),
    RequestFailed: Symbol('Speakers.NetworkManager.RequestFailed'),
    RequestFinished: Symbol('Speakers.NetworkManager.RequestFinished'),
  },

  FrameManager: {
    FrameAttached: Symbol('Speakers.FrameManager.FrameAttached'),
    FrameNavigated: Symbol('Speakers.FrameManager.FrameNavigated'),
    FrameDetached: Symbol('Speakers.FrameManager.FrameDetached'),
    LifecycleSpeaker: Symbol('Speakers.FrameManager.LifecycleSpeaker'),
    FrameNavigatedWithinDocument: Symbol('Speakers.FrameManager.FrameNavigatedWithinDocument'),
    ExecutionContextCreated: Symbol('Speakers.FrameManager.ExecutionContextCreated'),
    ExecutionContextDestroyed: Symbol('Speakers.FrameManager.ExecutionContextDestroyed'),
  },

  Connection: {
    Disconnected: Symbol('Speakers.Connection.Disconnected'),
  },

  CDPSession: {
    Disconnected: Symbol('Speakers.CDPSession.Disconnected'),
  },
};

module.exports = { Speakers };
